routing/localchans/manager.go

```diff
@@ -76,20 +76,34 @@ func (r *Manager) UpdatePolicy(newSchema routing.ChannelPolicy,
 			return nil
 		}
 
+		// Add edge to a list. We can't yet apply the new policy,
+		// because the updateEdge method will open a new db transaction.
+		edges = append(edges, discovery.EdgeWithInfo{
```

Comment:
I think `edges` should be initialized on first access. We can't be sure if `ForAllOutgoingChannels` is ever retried. Even though all of this is going to be ram cached soon.

Reply:
Hmm yes interesting. But it can't be fixed here I think? It seems that the deeper graph method needs a `reset` function passed in as well. In fact every graph method that accepts a callback needs a `reset` function?

There may be non-graph db methods too with a callback. We are accidentally stumbling into this, but perhaps there is more unsafe usage of external variables. The pattern to store items in an external list is quite common.

I do want to re-emphasize my earlier comment https://github.com/lightningnetwork/lnd/pull/5484#discussion_r674606066. I still think that switching to the serializable locking model is a risk that doesn't need to be taken at this time.

Reply:
I'd just add a flag and check it? If false => create the array, set flag true

Reply:
> Hmm yes interesting. But it can't be fixed here I think? It seems that the deeper graph method needs a `reset` function passed in as well. In fact every graph method that accepts a callback needs a `reset` function?
> 
> There may be non-graph db methods too with a callback. We are accidentally stumbling into this, but perhaps there is more unsafe usage of external variables. The pattern to store items in an external list is quite common.
> 
> I do want to re-emphasize my earlier comments [btcsuite/btcwallet#757 (comment)](https://github.com/btcsuite/btcwallet/pull/757#discussion_r673155676) and [#5484 (comment)](https://github.com/lightningnetwork/lnd/pull/5484#discussion_r674606066). I still think that switching to the serializable locking model is a risk that doesn't need to be taken at this time.

Also re: accidentally stumbling: I think during the fully remote PR we discussed that we need to make sure all DB transactions are in their closed forms and retries are handled. It may have been forgotten somehow as there was too many other threads of dicussions. Maybe worth merging this PR with https://github.com/lightningnetwork/lnd/pull/5642 and doing these fixes there.

Reply:
@guggero what do you think? I am not sure if it is a good idea to leak details of the database locking model into the graph interface.

Is there an issue in the repo to check whether all transactions are in their closed forms? I had an idea that could help with that: modify the code to artificially always run the code inside each transaction twice. It is then as if everything needs a retry. Itests, bench, etc, should still work.

Reply:
>Add a tx kvdb.RTx parameter to the FetchChannel method of the channel DB.

That is a great suggestion. The whole thing will be retried then with a fresh slice.

>What I don't understand is: Why did this never lead to a deadlock with bbolt?

Good question. I will look into it.

Reply:
That the deadlock doesn't occur with bbolt almost makes me think that somehow two different locks are being used. As if there are two instances of something.

Reply:
I am now looking at indeed the type of lock that postgres uses. Maybe you are right @guggero that it isn't exactly the same.

Reply:
Explanation why getting an `RLock` twice can lead to a deadlock: https://github.com/sasha-s/go-deadlock

Reply:
The same problem exists for bbolt. It just doesn't show up in the itest because timing is different. For a recursive read lock to go wrong, there needs to be a write that comes in between. By adding a `sleep(5)` inside the `ForAllOutgoingChannels` callback, the issue reproduces with bbolt on master.